// @ngx-formly/material/bundles/ngx-formly-material-checkbox.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/forms'), require('rxjs'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/material/checkbox')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/checkbox', ['exports', '@angular/core', '@ngx-formly/material/form-field', '@angular/forms', 'rxjs', '@angular/common', '@ngx-formly/core', '@angular/material/checkbox'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.checkbox = {}),global.ng.core,global['ngx-formly'].material['form-field'],global.ng.forms,global.rxjs,global.ng.common,null,global.ng.material.checkbox));
}(this, (function (exports,core,formField,forms,rxjs,common,core$1,checkbox) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldCheckbox = (function (_super) {
        __extends(FormlyFieldCheckbox, _super);
        function FormlyFieldCheckbox() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @param {?} $event
         * @return {?}
         */
        FormlyFieldCheckbox.prototype.change = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (this.to.change) {
                    this.to.change(this.field, $event);
                }
            };
        FormlyFieldCheckbox.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-checkbox',
                        template: "\n    <mat-checkbox\n      [formControl]=\"formControl\"\n      [id]=\"id\"\n      [formlyAttributes]=\"field\"\n      (change)=\"change($event)\"\n      [indeterminate]=\"to.indeterminate && model[key] === undefined\"\n      [color]=\"to.color\"\n      [labelPosition]=\"to.align || to.labelPosition\">\n      {{ to.label }}\n      <ng-container *ngIf=\"to.required && to.hideRequiredMarker !== true\">*</ng-container>\n    </mat-checkbox>\n  ",
                    },] },
        ];
        return FormlyFieldCheckbox;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldMultiCheckbox = (function (_super) {
        __extends(FormlyFieldMultiCheckbox, _super);
        function FormlyFieldMultiCheckbox() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @param {?} model
         * @param {?} field
         * @return {?}
         */
        FormlyFieldMultiCheckbox.createControl = /**
         * @param {?} model
         * @param {?} field
         * @return {?}
         */
            function (model, field) {
                if (!(field.templateOptions.options instanceof rxjs.Observable)) {
                    var /** @type {?} */ controlGroupConfig = field.templateOptions.options.reduce(function (previous, option) {
                        previous[option.key] = new forms.FormControl(model ? model[option.key] : undefined);
                        return previous;
                    }, {});
                    return new forms.FormGroup(controlGroupConfig, field.validators ? field.validators.validation : undefined, field.asyncValidators ? field.asyncValidators.validation : undefined);
                }
                else {
                    throw new Error("[Formly Error] You cannot pass an Observable to a multicheckbox yet.");
                }
            };
        FormlyFieldMultiCheckbox.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-multicheckbox',
                        template: "\n    <ng-container *ngFor=\"let option of to.options; let i = index;\">\n      <mat-checkbox [id]=\"id + '_' + i\"\n        [formControl]=\"formControl.get(option.key)\"\n        [formlyAttributes]=\"field\"\n        [color]=\"to.color\"\n        [labelPosition]=\"to.labelPosition\">\n          {{ option.value }}\n      </mat-checkbox>\n    </ng-container>\n  ",
                    },] },
        ];
        return FormlyFieldMultiCheckbox;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatCheckboxModule = (function () {
        function FormlyMatCheckboxModule() {
        }
        FormlyMatCheckboxModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyFieldCheckbox, FormlyFieldMultiCheckbox],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            checkbox.MatCheckboxModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [
                                    {
                                        name: 'checkbox',
                                        component: FormlyFieldCheckbox,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                hideFieldUnderline: true,
                                                indeterminate: true,
                                                floatLabel: 'always',
                                                hideLabel: true,
                                                align: 'start',
                                            },
                                        },
                                    },
                                    {
                                        name: 'multicheckbox',
                                        component: FormlyFieldMultiCheckbox,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                hideFieldUnderline: true,
                                                floatLabel: 'always',
                                                options: [],
                                            },
                                        },
                                    },
                                ],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatCheckboxModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatCheckboxModule = FormlyMatCheckboxModule;
    exports.FormlyFieldCheckbox = FormlyFieldCheckbox;
    exports.FormlyFieldMultiCheckbox = FormlyFieldMultiCheckbox;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
