// @ngx-formly/material/bundles/ngx-formly-material-select.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/select'), require('@ngx-formly/material/form-field'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/select', ['exports', '@angular/core', '@angular/material/select', '@ngx-formly/material/form-field', 'rxjs', 'rxjs/operators', '@angular/common', '@ngx-formly/core', '@angular/forms'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.select = {}),global.ng.core,global.ng.material.select,global['ngx-formly'].material['form-field'],global.rxjs,global.rxjs.operators,global.ng.common,null,global.ng.forms));
}(this, (function (exports,core,select,formField,rxjs,operators,common,core$1,forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldSelect = (function (_super) {
        __extends(FormlyFieldSelect, _super);
        function FormlyFieldSelect() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        Object.defineProperty(FormlyFieldSelect.prototype, "labelProp", {
            get: /**
             * @return {?}
             */ function () { return this.to["labelProp"] || 'label'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormlyFieldSelect.prototype, "valueProp", {
            get: /**
             * @return {?}
             */ function () { return this.to["valueProp"] || 'value'; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(FormlyFieldSelect.prototype, "groupProp", {
            get: /**
             * @return {?}
             */ function () { return this.to["groupProp"] || 'group'; },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} $event
         * @return {?}
         */
        FormlyFieldSelect.prototype.change = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (this.to.change) {
                    this.to.change(this.field, $event);
                }
            };
        FormlyFieldSelect.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-select',
                        template: "\n    <mat-select [id]=\"id\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\"\n      [multiple]=\"to.multiple\"\n      (selectionChange)=\"change($event)\"\n      [errorStateMatcher]=\"errorStateMatcher\">\n      <ng-container *ngFor=\"let item of to.options | formlySelectOptions:groupProp | async\">\n        <mat-optgroup *ngIf=\"item.group\" [label]=\"item.label\">\n          <mat-option *ngFor=\"let child of item.group\" [value]=\"child[valueProp]\" [disabled]=\"child.disabled\">\n            {{ child[labelProp] }}\n          </mat-option>\n        </mat-optgroup>\n        <mat-option *ngIf=\"!item.group\" [value]=\"item[valueProp]\" [disabled]=\"item.disabled\">{{ item[labelProp] }}</mat-option>\n      </ng-container>\n    </mat-select>\n  ",
                    },] },
        ];
        FormlyFieldSelect.propDecorators = {
            formFieldControl: [{ type: core.ViewChild, args: [select.MatSelect,] }]
        };
        return FormlyFieldSelect;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlySelectOptionsPipe = (function () {
        function FormlySelectOptionsPipe() {
        }
        /**
         * @param {?} options
         * @param {?=} groupBy
         * @return {?}
         */
        FormlySelectOptionsPipe.prototype.transform = /**
         * @param {?} options
         * @param {?=} groupBy
         * @return {?}
         */
            function (options, groupBy) {
                var _this = this;
                if (groupBy === void 0) {
                    groupBy = 'group';
                }
                if (!(options instanceof rxjs.Observable)) {
                    options = rxjs.of(options);
                }
                return ((options)).pipe(operators.map(function (value) { return _this.toOptions(value, groupBy); }));
            };
        /**
         * @param {?} options
         * @param {?} groupBy
         * @return {?}
         */
        FormlySelectOptionsPipe.prototype.toOptions = /**
         * @param {?} options
         * @param {?} groupBy
         * @return {?}
         */
            function (options, groupBy) {
                var /** @type {?} */ gOptions = [], /** @type {?} */ groups = {};
                options.map(function (option) {
                    if (!option[groupBy]) {
                        gOptions.push(option);
                    }
                    else {
                        if (groups[option[groupBy]]) {
                            groups[option[groupBy]].push(option);
                        }
                        else {
                            groups[option[groupBy]] = [option];
                            gOptions.push({
                                label: option[groupBy],
                                group: groups[option[groupBy]],
                            });
                        }
                    }
                });
                return gOptions;
            };
        FormlySelectOptionsPipe.decorators = [
            { type: core.Pipe, args: [{ name: 'formlySelectOptions' },] },
        ];
        return FormlySelectOptionsPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatSelectModule = (function () {
        function FormlyMatSelectModule() {
        }
        FormlyMatSelectModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyFieldSelect, FormlySelectOptionsPipe],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            select.MatSelectModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [{
                                        name: 'select',
                                        component: FormlyFieldSelect,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                options: [],
                                            },
                                        },
                                    }],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatSelectModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatSelectModule = FormlyMatSelectModule;
    exports.FormlyFieldSelect = FormlyFieldSelect;
    exports.ɵa = FormlySelectOptionsPipe;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
