// @ngx-formly/material/bundles/ngx-formly-material-toggle.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms'), require('@angular/material/slide-toggle')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/toggle', ['exports', '@angular/core', '@ngx-formly/material/form-field', '@angular/common', '@ngx-formly/core', '@angular/forms', '@angular/material/slide-toggle'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.toggle = {}),global.ng.core,global['ngx-formly'].material['form-field'],global.ng.common,null,global.ng.forms,global.ng.material['slide-toggle']));
}(this, (function (exports,core,formField,common,core$1,forms,slideToggle) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyToggleTypeComponent = (function (_super) {
        __extends(FormlyToggleTypeComponent, _super);
        function FormlyToggleTypeComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyToggleTypeComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-toggle',
                        template: "\n    <mat-slide-toggle [formControl]=\"formControl\" [formlyAttributes]=\"field\">\n      {{ to.label }}\n    </mat-slide-toggle>\n  ",
                    },] },
        ];
        return FormlyToggleTypeComponent;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatToggleModule = (function () {
        function FormlyMatToggleModule() {
        }
        FormlyMatToggleModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyToggleTypeComponent],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            slideToggle.MatSlideToggleModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [{
                                        name: 'toggle',
                                        component: FormlyToggleTypeComponent,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                hideFieldUnderline: true,
                                                floatLabel: 'always',
                                                hideLabel: true,
                                            },
                                        },
                                    }],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatToggleModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatToggleModule = FormlyMatToggleModule;
    exports.ɵa = FormlyToggleTypeComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
