// @ngx-formly/material/bundles/ngx-formly-material-datepicker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/material/input'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms'), require('@angular/material/datepicker')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/datepicker', ['exports', '@angular/core', '@ngx-formly/material/form-field', '@angular/material/input', '@angular/common', '@ngx-formly/core', '@angular/forms', '@angular/material/datepicker'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.datepicker = {}),global.ng.core,global['ngx-formly'].material['form-field'],global.ng.material.input,global.ng.common,null,global.ng.forms,global.ng.material.datepicker));
}(this, (function (exports,core,formField,input,common,core$1,forms,datepicker) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyDatepickerTypeComponent = (function (_super) {
        __extends(FormlyDatepickerTypeComponent, _super);
        function FormlyDatepickerTypeComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyDatepickerTypeComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-datepicker',
                        template: "\n    <input matInput\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [formControl]=\"formControl\"\n      [matDatepicker]=\"picker\"\n      [matDatepickerFilter]=\"to.datepickerOptions.filter\"\n      [max]=\"to.datepickerOptions.max\"\n      [min]=\"to.datepickerOptions.min\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\">\n    <ng-template #matSuffix>\n      <mat-datepicker-toggle [for]=\"picker\"></mat-datepicker-toggle>\n    </ng-template>\n    <mat-datepicker #picker\n      [color]=\"to.color\"\n      [touchUi]=\"to.datepickerOptions.touchUi\"\n      [startView]=\"to.datepickerOptions.startView\"\n      [startAt]=\"to.datepickerOptions.startAt\">\n    </mat-datepicker>\n  ",
                    },] },
        ];
        FormlyDatepickerTypeComponent.propDecorators = {
            formFieldControl: [{ type: core.ViewChild, args: [input.MatInput,] }]
        };
        return FormlyDatepickerTypeComponent;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatDatepickerModule = (function () {
        function FormlyMatDatepickerModule() {
        }
        FormlyMatDatepickerModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyDatepickerTypeComponent],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            input.MatInputModule,
                            datepicker.MatDatepickerModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [{
                                        name: 'datepicker',
                                        component: FormlyDatepickerTypeComponent,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                datepickerOptions: {
                                                    startView: 'month',
                                                },
                                            },
                                        },
                                    }],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatDatepickerModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatDatepickerModule = FormlyMatDatepickerModule;
    exports.ɵa = FormlyDatepickerTypeComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
