// @ngx-formly/material/bundles/ngx-formly-material-radio.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@ngx-formly/material/form-field'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms'), require('@angular/material/radio')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/radio', ['exports', '@angular/core', '@ngx-formly/material/form-field', '@angular/common', '@ngx-formly/core', '@angular/forms', '@angular/material/radio'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.radio = {}),global.ng.core,global['ngx-formly'].material['form-field'],global.ng.common,null,global.ng.forms,global.ng.material.radio));
}(this, (function (exports,core,formField,common,core$1,forms,radio) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldRadio = (function (_super) {
        __extends(FormlyFieldRadio, _super);
        function FormlyFieldRadio() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @param {?} $event
         * @return {?}
         */
        FormlyFieldRadio.prototype.change = /**
         * @param {?} $event
         * @return {?}
         */
            function ($event) {
                if (this.to.change) {
                    this.to.change(this.field, $event);
                }
            };
        FormlyFieldRadio.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-radio',
                        template: "\n    <mat-radio-group [formControl]=\"formControl\" [formlyAttributes]=\"field\" (change)=\"change($event)\">\n      <mat-radio-button *ngFor=\"let option of to.options; let i = index;\"\n        [id]=\"id + '_' + i\"\n        [color]=\"to.color\"\n        [labelPosition]=\"to.labelPosition\"\n        [value]=\"option.label ? option.value : option.key\">\n        {{ option.label ? option.label : option.value }}\n      </mat-radio-button>\n    </mat-radio-group>\n  ",
                    },] },
        ];
        return FormlyFieldRadio;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatRadioModule = (function () {
        function FormlyMatRadioModule() {
        }
        FormlyMatRadioModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyFieldRadio],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            radio.MatRadioModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [{
                                        name: 'radio',
                                        component: FormlyFieldRadio,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                hideFieldUnderline: true,
                                                floatLabel: 'always',
                                                options: [],
                                            },
                                        },
                                    }],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatRadioModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatRadioModule = FormlyMatRadioModule;
    exports.FormlyFieldRadio = FormlyFieldRadio;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
