// @ngx-formly/material/bundles/ngx-formly-material-input.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/input'), require('@ngx-formly/material/form-field'), require('@angular/common'), require('@ngx-formly/core'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@ngx-formly/material/input', ['exports', '@angular/core', '@angular/material/input', '@ngx-formly/material/form-field', '@angular/common', '@ngx-formly/core', '@angular/forms'], factory) :
    (factory((global['ngx-formly'] = global['ngx-formly'] || {}, global['ngx-formly'].material = global['ngx-formly'].material || {}, global['ngx-formly'].material.input = {}),global.ng.core,global.ng.material.input,global['ngx-formly'].material['form-field'],global.ng.common,null,global.ng.forms));
}(this, (function (exports,core,input,formField,common,core$1,forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b)
            if (b.hasOwnProperty(p))
                d[p] = b[p]; };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldInput = (function (_super) {
        __extends(FormlyFieldInput, _super);
        function FormlyFieldInput() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFieldInput.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-input',
                        template: "\n    <input matInput\n      [id]=\"id\"\n      [type]=\"to.type || 'text'\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\">\n  ",
                    },] },
        ];
        FormlyFieldInput.propDecorators = {
            formFieldControl: [{ type: core.ViewChild, args: [input.MatInput,] }]
        };
        return FormlyFieldInput;
    }(formField.FieldType));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyFieldTextArea = (function (_super) {
        __extends(FormlyFieldTextArea, _super);
        function FormlyFieldTextArea() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        FormlyFieldTextArea.decorators = [
            { type: core.Component, args: [{
                        selector: 'formly-field-mat-textarea',
                        template: "\n    <textarea matInput\n      [id]=\"id\"\n      [name]=\"key\"\n      [formControl]=\"formControl\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [cols]=\"to.cols\"\n      [rows]=\"to.rows\"\n      [placeholder]=\"to.placeholder\"\n      [formlyAttributes]=\"field\">\n    </textarea>\n  ",
                    },] },
        ];
        return FormlyFieldTextArea;
    }(FormlyFieldInput));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var FormlyMatInputModule = (function () {
        function FormlyMatInputModule() {
        }
        FormlyMatInputModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [FormlyFieldInput, FormlyFieldTextArea],
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            input.MatInputModule,
                            formField.FormlyMatFormFieldModule,
                            core$1.FormlyModule.forChild({
                                types: [
                                    {
                                        name: 'input',
                                        component: FormlyFieldInput,
                                        wrappers: ['form-field'],
                                    },
                                    {
                                        name: 'textarea',
                                        component: FormlyFieldTextArea,
                                        wrappers: ['form-field'],
                                        defaultOptions: {
                                            templateOptions: {
                                                cols: 1,
                                                rows: 1,
                                            },
                                        },
                                    },
                                ],
                            }),
                        ],
                    },] },
        ];
        return FormlyMatInputModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.FormlyMatInputModule = FormlyMatInputModule;
    exports.FormlyFieldInput = FormlyFieldInput;
    exports.FormlyFieldTextArea = FormlyFieldTextArea;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
