// angular-formio/resource/formio-resource.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { FormioResourceConfig, FormioResources, FormioResourceService, FormioResourceComponent, FormioResourceViewComponent, FormioResourceEditComponent, FormioResourceDeleteComponent, FormioResourceCreateComponent, FormioResourceIndexComponent, FormioResourceRoutes, FormioResource } from './index';
export { FormioAlerts as ɵh } from '../components/alerts/formio.alerts';
export { FormioAlertsComponent as ɵg } from '../components/alerts/formio.alerts.component';
export { FormBuilderComponent as ɵe } from '../components/formbuilder/formbuilder.component';
export { FormioComponent as ɵd } from '../components/formio/formio.component';
export { FormioLoader as ɵb } from '../components/loader/formio.loader';
export { FormioLoaderComponent as ɵf } from '../components/loader/formio.loader.component';
export { FormioAppConfig as ɵa } from '../formio.config';
export { FormioModule as ɵc } from '../formio.module';
export { FormioGridComponent as ɵj } from '../grid/grid.component';
export { FormioGrid as ɵi } from '../grid/grid.module';
