// angular-formio/resource/resources.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
/**
 * @record
 */
export function FormioResourceMap() { }
function FormioResourceMap_tsickle_Closure_declarations() {
    /* TODO: handle strange member:
    [name: string]: any;
    */
}
var FormioResources = /** @class */ (function () {
    function FormioResources() {
        this.resources = {};
        this.error = new EventEmitter();
        this.onError = this.error;
        this.resources = {};
    }
    FormioResources.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    FormioResources.ctorParameters = function () { return []; };
    return FormioResources;
}());
export { FormioResources };
function FormioResources_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    FormioResources.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    FormioResources.ctorParameters;
    /** @type {?} */
    FormioResources.prototype.resources;
    /** @type {?} */
    FormioResources.prototype.error;
    /** @type {?} */
    FormioResources.prototype.onError;
}
