// angular-formio/auth/auth.config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
/**
 * @record
 */
export function FormioAuthFormConfig() { }
function FormioAuthFormConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    FormioAuthFormConfig.prototype.path;
    /** @type {?|undefined} */
    FormioAuthFormConfig.prototype.form;
    /** @type {?|undefined} */
    FormioAuthFormConfig.prototype.component;
}
/**
 * @record
 */
export function FormioAuthRouteConfig() { }
function FormioAuthRouteConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    FormioAuthRouteConfig.prototype.auth;
    /** @type {?|undefined} */
    FormioAuthRouteConfig.prototype.login;
    /** @type {?|undefined} */
    FormioAuthRouteConfig.prototype.register;
}
var FormioAuthConfig = /** @class */ (function () {
    function FormioAuthConfig() {
    }
    FormioAuthConfig.decorators = [
        { type: Injectable },
    ];
    return FormioAuthConfig;
}());
export { FormioAuthConfig };
function FormioAuthConfig_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    FormioAuthConfig.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    FormioAuthConfig.ctorParameters;
    /** @type {?} */
    FormioAuthConfig.prototype.component;
    /** @type {?} */
    FormioAuthConfig.prototype.login;
    /** @type {?} */
    FormioAuthConfig.prototype.register;
}
