// angular-formio/auth/auth.routes.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { FormioAuthComponent } from './auth.component';
import { FormioAuthLoginComponent } from './login/login.component';
import { FormioAuthRegisterComponent } from './register/register.component';
/**
 * @param {?=} config
 * @return {?}
 */
export function FormioAuthRoutes(config) {
    config = config || {};
    return [
        {
            path: '',
            component: config.auth || FormioAuthComponent,
            children: [
                {
                    path: '',
                    redirectTo: 'login',
                    pathMatch: 'full'
                },
                {
                    path: 'login',
                    component: config.login || FormioAuthLoginComponent
                },
                {
                    path: 'register',
                    component: config.register || FormioAuthRegisterComponent
                }
            ]
        }
    ];
}
;
