// @ngx-formly/material/esm2015/checkbox/checkbox.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { MatCheckboxModule } from '@angular/material/checkbox';
import { FormlyFieldCheckbox } from './checkbox.type';
import { FormlyFieldMultiCheckbox } from './multicheckbox.type';
export class FormlyMatCheckboxModule {
}
FormlyMatCheckboxModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyFieldCheckbox, FormlyFieldMultiCheckbox],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatCheckboxModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [
                            {
                                name: 'checkbox',
                                component: FormlyFieldCheckbox,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        hideFieldUnderline: true,
                                        indeterminate: true,
                                        floatLabel: 'always',
                                        hideLabel: true,
                                        align: 'start',
                                    },
                                },
                            },
                            {
                                name: 'multicheckbox',
                                component: FormlyFieldMultiCheckbox,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        hideFieldUnderline: true,
                                        floatLabel: 'always',
                                        options: [],
                                    },
                                },
                            },
                        ],
                    }),
                ],
            },] },
];
