// @ngx-formly/material/esm5/slider/slider.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatSliderModule } from '@angular/material/slider';
import { FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { FormlySliderTypeComponent } from './slider.type';
var FormlyMatSliderModule = /** @class */ (function () {
    function FormlyMatSliderModule() {
    }
    FormlyMatSliderModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FormlySliderTypeComponent],
                    imports: [
                        CommonModule,
                        ReactiveFormsModule,
                        MatSliderModule,
                        FormlyMatFormFieldModule,
                        FormlyModule.forChild({
                            types: [{
                                    name: 'slider',
                                    component: FormlySliderTypeComponent,
                                    wrappers: ['form-field'],
                                    defaultOptions: {
                                        templateOptions: {
                                            hideFieldUnderline: true,
                                            floatLabel: 'always',
                                        },
                                    },
                                }],
                        }),
                    ],
                },] },
    ];
    return FormlyMatSliderModule;
}());
export { FormlyMatSliderModule };
