// @ngx-formly/material/esm5/ngx-formly-material-radio.js
import { __extends } from 'tslib';
import { Component, NgModule } from '@angular/core';
import { FieldType, FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatRadioModule } from '@angular/material/radio';

var FormlyFieldRadio = /** @class */ (function (_super) {
    __extends(FormlyFieldRadio, _super);
    function FormlyFieldRadio() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FormlyFieldRadio.prototype.change = function ($event) {
        if (this.to.change) {
            this.to.change(this.field, $event);
        }
    };
    return FormlyFieldRadio;
}(FieldType));
FormlyFieldRadio.decorators = [
    { type: Component, args: [{
                selector: 'formly-field-mat-radio',
                template: "\n    <mat-radio-group [formControl]=\"formControl\" [formlyAttributes]=\"field\" (change)=\"change($event)\">\n      <mat-radio-button *ngFor=\"let option of to.options; let i = index;\"\n        [id]=\"id + '_' + i\"\n        [color]=\"to.color\"\n        [labelPosition]=\"to.labelPosition\"\n        [value]=\"option.label ? option.value : option.key\">\n        {{ option.label ? option.label : option.value }}\n      </mat-radio-button>\n    </mat-radio-group>\n  ",
            },] },
];
var FormlyMatRadioModule = /** @class */ (function () {
    function FormlyMatRadioModule() {
    }
    return FormlyMatRadioModule;
}());
FormlyMatRadioModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyFieldRadio],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatRadioModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [{
                                name: 'radio',
                                component: FormlyFieldRadio,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        hideFieldUnderline: true,
                                        floatLabel: 'always',
                                        options: [],
                                    },
                                },
                            }],
                    }),
                ],
            },] },
];

export { FormlyMatRadioModule, FormlyFieldRadio };
