// @ngx-formly/material/esm5/ngx-formly-material-toggle.js
import { __extends } from 'tslib';
import { Component, NgModule } from '@angular/core';
import { FieldType, FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';

var FormlyToggleTypeComponent = /** @class */ (function (_super) {
    __extends(FormlyToggleTypeComponent, _super);
    function FormlyToggleTypeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormlyToggleTypeComponent;
}(FieldType));
FormlyToggleTypeComponent.decorators = [
    { type: Component, args: [{
                selector: 'formly-field-mat-toggle',
                template: "\n    <mat-slide-toggle [formControl]=\"formControl\" [formlyAttributes]=\"field\">\n      {{ to.label }}\n    </mat-slide-toggle>\n  ",
            },] },
];
var FormlyMatToggleModule = /** @class */ (function () {
    function FormlyMatToggleModule() {
    }
    return FormlyMatToggleModule;
}());
FormlyMatToggleModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyToggleTypeComponent],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatSlideToggleModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [{
                                name: 'toggle',
                                component: FormlyToggleTypeComponent,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        hideFieldUnderline: true,
                                        floatLabel: 'always',
                                        hideLabel: true,
                                    },
                                },
                            }],
                    }),
                ],
            },] },
];

export { FormlyMatToggleModule, FormlyToggleTypeComponent as ɵa };
