// @ngx-formly/material/esm5/ngx-formly-material-input.js
import { __extends } from 'tslib';
import { Component, ViewChild, NgModule } from '@angular/core';
import { MatInput, MatInputModule } from '@angular/material/input';
import { FieldType, FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';

var FormlyFieldInput = /** @class */ (function (_super) {
    __extends(FormlyFieldInput, _super);
    function FormlyFieldInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormlyFieldInput;
}(FieldType));
FormlyFieldInput.decorators = [
    { type: Component, args: [{
                selector: 'formly-field-mat-input',
                template: "\n    <input matInput\n      [id]=\"id\"\n      [type]=\"to.type || 'text'\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      [placeholder]=\"to.placeholder\">\n  ",
            },] },
];
FormlyFieldInput.propDecorators = {
    "formFieldControl": [{ type: ViewChild, args: [MatInput,] },],
};
var FormlyFieldTextArea = /** @class */ (function (_super) {
    __extends(FormlyFieldTextArea, _super);
    function FormlyFieldTextArea() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormlyFieldTextArea;
}(FormlyFieldInput));
FormlyFieldTextArea.decorators = [
    { type: Component, args: [{
                selector: 'formly-field-mat-textarea',
                template: "\n    <textarea matInput\n      [id]=\"id\"\n      [name]=\"key\"\n      [formControl]=\"formControl\"\n      [errorStateMatcher]=\"errorStateMatcher\"\n      [cols]=\"to.cols\"\n      [rows]=\"to.rows\"\n      [placeholder]=\"to.placeholder\"\n      [formlyAttributes]=\"field\">\n    </textarea>\n  ",
            },] },
];
var FormlyMatInputModule = /** @class */ (function () {
    function FormlyMatInputModule() {
    }
    return FormlyMatInputModule;
}());
FormlyMatInputModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyFieldInput, FormlyFieldTextArea],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatInputModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [
                            {
                                name: 'input',
                                component: FormlyFieldInput,
                                wrappers: ['form-field'],
                            },
                            {
                                name: 'textarea',
                                component: FormlyFieldTextArea,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        cols: 1,
                                        rows: 1,
                                    },
                                },
                            },
                        ],
                    }),
                ],
            },] },
];

export { FormlyMatInputModule, FormlyFieldInput, FormlyFieldTextArea };
