// @ngx-formly/material/esm5/ngx-formly-material-slider.js
import { __extends } from 'tslib';
import { Component, NgModule } from '@angular/core';
import { FieldType, FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { MatSliderModule } from '@angular/material/slider';

var FormlySliderTypeComponent = /** @class */ (function (_super) {
    __extends(FormlySliderTypeComponent, _super);
    function FormlySliderTypeComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormlySliderTypeComponent;
}(FieldType));
FormlySliderTypeComponent.decorators = [
    { type: Component, args: [{
                selector: 'formly-field-mat-slider',
                template: "\n    <mat-slider\n      [style.width]=\"'100%'\"\n      [formControl]=\"formControl\"\n      [formlyAttributes]=\"field\"\n      [color]=\"to.color\">\n    </mat-slider>\n  ",
            },] },
];
var FormlyMatSliderModule = /** @class */ (function () {
    function FormlyMatSliderModule() {
    }
    return FormlyMatSliderModule;
}());
FormlyMatSliderModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlySliderTypeComponent],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatSliderModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [{
                                name: 'slider',
                                component: FormlySliderTypeComponent,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        hideFieldUnderline: true,
                                        floatLabel: 'always',
                                    },
                                },
                            }],
                    }),
                ],
            },] },
];

export { FormlyMatSliderModule, FormlySliderTypeComponent as ɵa };
