// angular-formio/resource/index/index.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component } from '@angular/core';
import { Router, ActivatedRoute } from '@angular/router';
import { FormioResourceService } from '../resource.service';
import { FormioResourceConfig } from '../resource.config';
import { each } from 'lodash';
var FormioResourceIndexComponent = /** @class */ (function () {
    function FormioResourceIndexComponent(service, route, router, config) {
        this.service = service;
        this.route = route;
        this.router = router;
        this.config = config;
    }
    /**
     * @return {?}
     */
    FormioResourceIndexComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.service.initialize();
        this.gridQuery = {};
        if (this.service &&
            this.service.onParents &&
            this.config.parents &&
            this.config.parents.length) {
            // Wait for the parents to load before loading this grid.
            this.service.onParents.subscribe(function (parents) {
                each(parents, function (parent) {
                    if (parent) {
                        _this.gridQuery['data.' + parent.name + '._id'] =
                            parent.resource._id;
                    }
                });
                // Set the source to load the grid.
                // Set the source to load the grid.
                _this.gridSrc = _this.service.formUrl;
            });
        }
        else if (this.service.formUrl) {
            this.gridSrc = this.service.formUrl;
        }
    };
    /**
     * @param {?} row
     * @return {?}
     */
    FormioResourceIndexComponent.prototype.onSelect = /**
     * @param {?} row
     * @return {?}
     */
    function (row) {
        this.router.navigate([row._id, 'view'], { relativeTo: this.route });
    };
    FormioResourceIndexComponent.decorators = [
        { type: Component, args: [{
                    template: "<formio-grid [src]=\"gridSrc\" [query]=\"gridQuery\" [onForm]=\"service.formLoaded\" (select)=\"onSelect($event)\" (error)=\"service.onError($event)\"></formio-grid> <button class=\"btn btn-primary\" *ngIf=\"service.form\" routerLink=\"new\"><span class=\"glyphicon glyphicon-plus\"></span> New {{ service.form.title }}</button> "
                },] },
    ];
    /** @nocollapse */
    FormioResourceIndexComponent.ctorParameters = function () { return [
        { type: FormioResourceService, },
        { type: ActivatedRoute, },
        { type: Router, },
        { type: FormioResourceConfig, },
    ]; };
    return FormioResourceIndexComponent;
}());
export { FormioResourceIndexComponent };
function FormioResourceIndexComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    FormioResourceIndexComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    FormioResourceIndexComponent.ctorParameters;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.gridSrc;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.gridQuery;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.service;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.route;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.router;
    /** @type {?} */
    FormioResourceIndexComponent.prototype.config;
}
