// @ngx-formly/material/esm2015/input/input.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormlyModule } from '@ngx-formly/core';
import { ReactiveFormsModule } from '@angular/forms';
import { FormlyMatFormFieldModule } from '@ngx-formly/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { FormlyFieldInput } from './input.type';
import { FormlyFieldTextArea } from './textarea.type';
export class FormlyMatInputModule {
}
FormlyMatInputModule.decorators = [
    { type: NgModule, args: [{
                declarations: [FormlyFieldInput, FormlyFieldTextArea],
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    MatInputModule,
                    FormlyMatFormFieldModule,
                    FormlyModule.forChild({
                        types: [
                            {
                                name: 'input',
                                component: FormlyFieldInput,
                                wrappers: ['form-field'],
                            },
                            {
                                name: 'textarea',
                                component: FormlyFieldTextArea,
                                wrappers: ['form-field'],
                                defaultOptions: {
                                    templateOptions: {
                                        cols: 1,
                                        rows: 1,
                                    },
                                },
                            },
                        ],
                    }),
                ],
            },] },
];
